macrosynergy/panel/linear_composite.py

```diff
@@ -1,173 +1,191 @@
-    df: pd.DataFrame,
-    xcats: Union[str, List[str]],
-    cids: Optional[List[str]] = None,
-    weights: Union[List[float], np.ndarray, pd.Series] = None,
-    signs: Union[List[float], np.ndarray, pd.Series] = None,
-    start: str = None,
-    end: str = None,
+
+def _calc(
+    data_df: pd.DataFrame,
+    weights_df: pd.DataFrame,
     normalize_weights: bool = True,
-    complete_xcats: bool = True,
-    new_xcat="NEW",
+    complete: bool = False,
+    mode: str = "xcat_agg",
 ):
-    if not len(xcats) == len(weights) == len(signs):
-        raise ValueError("xcats, weights, and signs must have same length")
-    if not np.isclose(np.sum(weights), 1) or normalize_weights:
-        if not normalize_weights:
-            warnings.warn("`weights` does not sum to 1 and will be normalized. w←w/∑w")
-        weights = weights / np.sum(weights)
-    if not np.all(np.isin(signs, [1, -1])):
-        warnings.warn("signs must be 1 or -1. They will be coerced to 1 or -1.")
-        signs = np.abs(signs) / signs  # should be faster?
-
-    # main function is here and below.
-    weights = pd.Series(weights * signs, index=xcats)
-
-    dfc: pd.DataFrame = reduce_df(df=df, cids=cids, xcats=xcats, start=start, end=end)
-
-    # dataframe with the xcats as columns and rows as cid-date combinations
-    dfc_wide = dfc.set_index(["cid", "real_date", "xcat"])["value"].unstack(level=2)
-    # dataframe for weights with same index as dfc_wide: each column will be a weight
-    weights_wide = pd.DataFrame(
-        data=[weights.sort_index()], index=dfc_wide.index, columns=dfc_wide.columns
-    )
-    # boolean mask to help us work out the calcs
-    mask = dfc_wide.isna()
-    # series with an index of dfc_wide, and a value equal to the sum of the weights
-    weights_sum = weights_wide[~mask].abs().sum(axis=1)
-    # re-weighting the weights to sum to 1 considering the available xcats
-    adj_weights_wide = weights_wide[~mask].div(weights_sum, axis=0)
-    # final single series: the linear combination of the xcats and the weights
-
-    out_df = (dfc_wide * adj_weights_wide).sum(axis=1)
-
-    if complete_xcats:
-        out_df[mask.any(axis=1)] = np.NaN
+    """Main calculation function for linear_composite()"""
+
+    # Create a boolean mask to help us work out the calcs
+    nan_mask: pd.DataFrame = data_df.isna() | weights_df.isna()
+
+    # Normalize weights (if requested)
+    if normalize_weights:
+        adj_weights_wide = weights_df[~nan_mask].div(
+            weights_df[~nan_mask].abs().sum(axis=1), axis=0
+        )
+        adj_weights_wide[nan_mask] = np.NaN
+
+        assert np.allclose(
+            adj_weights_wide.abs().sum(axis=1), 1
+        ), "Weights do not sum to 1. Normalization failed."
+
+    # Multiply the weights by the target data
+    out_df = data_df * adj_weights_wide
+
+    # Remove periods with missing data (if requested)
+    if complete:
+        out_df[nan_mask.any(axis=1)] = np.NaN
     else:
```

Comment:
lines 46-47 look redundant as missing rows imply that weights don't add to 1.

Reply:
Have changed it out to remove all rows with all nans, before we sum the columns